stdlib/experimental/dynamic/json.go

```diff
@@ -0,0 +1,229 @@
+	var i interface{}
+	err := json.Unmarshal(bytes, &i)
+	if err != nil {
+		return nil, err
+	}
+	inner, err := toValue(i)
+	if err != nil {
+		return nil, err
+	}
+	return values.NewDynamic(inner), nil
+}
+
+// toValue converts a Go value that can be produced by json.Unmarshal into its corresponding Flux value.
+func toValue(i interface{}) (values.Value, error) {
+	switch t := i.(type) {
+	case string:
+		return values.NewString(t), nil
+	case bool:
+		return values.NewBool(t), nil
+	case float64:
+		return values.NewFloat(t), nil
+	case []interface{}:
+		return toArray(t)
+	case map[string]interface{}:
+		return toObject(t)
+	}
+	if i == nil {
+		return values.Null, nil
+	}
+	return nil, errors.Newf(codes.Internal, "unsupported json type %T", i)
+}
+
+func toArray(a []interface{}) (values.Array, error) {
+	vals := make([]values.Value, len(a))
+	for i, v := range a {
+		val, err := toValue(v)
+		if err != nil {
+			return nil, err
+		}
+
+		vals[i] = values.NewDynamic(val)
+	}
+	arrayTyp := semantic.NewArrayType(semantic.NewDynamicType())
+	return values.NewArrayWithBacking(arrayTyp, vals), nil
+}
+
+func toObject(m map[string]interface{}) (values.Value, error) {
+	vals := make(map[string]values.Value, len(m))
+	for k, v := range m {
+		val, err := toValue(v)
+		if err != nil {
+			return nil, err
+		}
+		vals[k] = values.NewDynamic(val)
+	}
+	return values.NewObjectWithValues(vals), nil
+}
+
+var jsonEncode = values.NewFunction(
+	"encode",
+	runtime.MustLookupBuiltinType("experimental/dynamic", "jsonEncode"),
+	func(ctx context.Context, args values.Object) (values.Value, error) {
+		v, ok := args.Get("v")
+		if !ok {
+			return nil, errors.New(codes.Invalid, "missing parameter \"v\"")
+		}
+		val, err := convertValue(v)
+		if err != nil {
+			return nil, err
+		}
+		bytes, err := json.Marshal(val)
+		if err != nil {
+			return nil, err
+		}
+		return values.NewBytes(bytes), nil
+	},
+	false,
+)
+
+// convertValue takes a Flux value and converts it to Go native values
+// that Go's standard JSON serialization can work with.
+//
+// This code is only slightly modified from the non-experimental
+// version in stdlib/json/encode.go.
+func convertValue(v values.Value) (interface{}, error) {
+	if v.IsNull() {
+		return nil, nil
+	}
+	switch n := v.Type().Nature(); n {
+	case semantic.Dynamic:
+		return convertValue(v.Dynamic().Inner())
+	case semantic.String:
+		return v.Str(), nil
+	case semantic.Bytes:
+		return v.Bytes(), nil
+	case semantic.Int:
+		return v.Int(), nil
+	case semantic.UInt:
+		return v.UInt(), nil
+	case semantic.Float:
+		return v.Float(), nil
+	case semantic.Bool:
+		return v.Bool(), nil
+	case semantic.Time:
+		return v.Time().Time().Format(time.RFC3339Nano), nil
+	case semantic.Duration:
+		return v.Duration(), nil
+	case semantic.Regexp:
+		return v.Regexp().String(), nil
+	case semantic.Array:
+		arr := v.Array()
+		a := make([]interface{}, arr.Len())
+		var rangeErr error
+		arr.Range(func(i int, v values.Value) {
```

Comment:
I wondered whether I should use `Array.Range` or just use a plain loop while I was looking at `asArray`. I opted for "plain loop" just because it seemed more minimal. What made you prefer it?

My guess: in this function we're iterating over different collection types (array, dict, object) and the `Range()` method might make this work more uniform.

Reply:
This code I mostly copied from the non-dynamic implementation of `encode`. I only had to add the case to `semantic.Dynamic` above. So it was like this already. Looking at the alternative, I can't say that I have a strong preference.

I considered just changing the regular version of `json.encode` to be able to handle `dynamic`, since it's an additive change that shouldn't break with existing behavior, but decided it would be more prudent to copy it.